src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs

```diff
@@ -173,82 +176,105 @@ public void InitializeService(ServiceHost serviceHost)
+
+                await ConnectionService.Connect(connectParams);
+
+                ConnectionInfo newConn;
+                ConnectionService.TryFindConnection(randomUri, out newConn);
+
+                Func<string, Task> queryCreateFailureAction = message => requestContext.SendError(message);
+
+                ResultOnlyContext<SimpleExecuteResult> newContext = new ResultOnlyContext<SimpleExecuteResult>(requestContext);
+
+                // handle sending event back when the query completes
+                Query.QueryAsyncEventHandler queryComplete = async q =>
                 {
-                    RowCount = q.Batches[0].ResultSets[0].RowCount,
-                    ColumnInfo = q.Batches[0].ResultSets[0].Columns,
-                    Rows = subset.Rows
+                    try
+                    {
+                        // check to make sure any results were recieved
+                        if (q.Batches.Length == 0 || q.Batches[0].ResultSets.Count == 0) 
+                        {
+                            await requestContext.SendError(SR.QueryServiceResultSetHasNoResults);
+                            return;
+                        } 
+
+                        var rowCount = q.Batches[0].ResultSets[0].RowCount;
+                        // check to make sure there is a safe amount of rows to load into memory
+                        if (rowCount > Int32.MaxValue) 
+                        {
+                            await requestContext.SendError(SR.QueryServiceResultSetTooLarge);
+                            return;
+                        }
+                        
+                        SubsetParams subsetRequestParams = new SubsetParams
+                        {
+                            OwnerUri = randomUri,
+                            BatchIndex = 0,
+                            ResultSetIndex = 0,
+                            RowsStartIndex = 0,
+                            RowsCount = Convert.ToInt32(rowCount)
+                        };
+                        // get the data to send back
+                        ResultSetSubset subset = await InterServiceResultSubset(subsetRequestParams);
+                        SimpleExecuteResult result = new SimpleExecuteResult
+                        {
+                            RowCount = q.Batches[0].ResultSets[0].RowCount,
+                            ColumnInfo = q.Batches[0].ResultSets[0].Columns,
+                            Rows = subset.Rows
+                        };
+                        await requestContext.SendResult(result);
+                    } 
+                    finally 
+                    {
+                        Query removedQuery;
+                        // remove the active query since we are done with it
+                        ActiveQueries.TryRemove(randomUri, out removedQuery);
+                        ConnectionService.Disconnect(new DisconnectParams(){
```

Comment:
I think you'll want to disconnect in all scenarios, in other words, I'd move this to the outside try/catch/finally block.